uspace/lib/c/generic/adt/array.c

```diff
@@ -0,0 +1,192 @@
+ * array.h because of type genericity.
+ */
+
+#include <assert.h>
+#include <errno.h>
+#include <adt/array.h>
+#include <macros.h>
+#include <mem.h>
+#include <stdlib.h>
+
+static errno_t array_realloc(array_t *da, size_t capacity)
+{
+	if (capacity == da->capacity) {
+		return EOK;
+	}
+
+	void *new_data = realloc(da->_data, da->_item_size * capacity);
+	if (new_data || capacity == 0) {
+		da->_data = new_data;
+		da->capacity = capacity;
+	}
+	return (new_data == NULL && capacity != 0) ? ENOMEM : EOK;
+}
+
+void array_destroy(array_t *da)
+{
+	array_clear(da);
+	free(da->_data);
+	da->capacity = 0;
+}
+
+/** Remove item at given position, shift rest of array */
+void array_remove(array_t *da, size_t index)
+{
+	assert(index < da->size);
+	_array_unshift(da, index, 1);
+	errno_t rc = array_reserve(da, da->size);
+	assert(rc == EOK);
+}
+
+/** Clear dynamic array (empty) */
+void array_clear(array_t *da)
+{
+	da->size = 0;
+}
+
+/** Clear subsequence of array
+ *
+ * @param[in/out]  da
+ * @param[in]      begin  index of first item to remove
+ * @param[in]      end    index behind last item to remove
+ */
+void array_clear_range(array_t *da, size_t begin, size_t end)
+{
+	assert(begin < da->size);
+	assert(end <= da->size);
+
+	_array_unshift(da, begin, end - begin);
+	errno_t rc = array_reserve(da, da->size);
+	assert(rc == EOK);
+}
+
+/** Concatenate two arrays
+ *
+ * @param[in/out]  da1  first array and concatenated output
+ * @param[in]      da2  second array, it is untouched
+ *
+ * @return EOK on success
+ * @return ENOMEM when allocation fails
+ */
+errno_t array_concat(array_t *da1, array_t *da2)
+{
+	assert(da1->_item_size == da2->_item_size);
+
+	errno_t rc = array_reserve(da1, da1->size + da2->size);
+	if (rc != EOK) {
+		return rc;
+	}
+	void *dst = da1->_data + da1->size * da1->_item_size;
+	void *src = da2->_data;
+	size_t size = da1->_item_size * da2->size;
+	memcpy(dst, src, size);
+	da1->size += da2->size;
+
+	return EOK;
+}
+
+/** Grows/shrinks array so that it effeciently stores desired capacity
+ *
+ * @param      da
+ * @param[in]  desired capacity of array (items)
+ *
+ * @return EOK
+ * @return ENOMEM
+ */
+errno_t array_reserve(array_t *da, size_t capacity)
+{
+	const size_t factor = 2;
+	size_t new_capacity;
+	if (capacity > da->capacity) {
+		new_capacity = max(da->capacity * factor, capacity);
+	} else if (capacity < da->capacity / factor) {
+		new_capacity = min(da->capacity / factor, capacity);
+	} else {
+		new_capacity = capacity;
+	}
+
+	return array_realloc(da, new_capacity);
+}
+
+void _array_initialize(array_t *da, size_t item_size)
+{
+	da->_item_size = item_size;
+	da->_data = NULL;
+
+	da->capacity = 0;
+	da->size = 0;
+}
+
+/** Shift block of array
+ *
+ * Extends size of dynamic array, assumes sufficient capacity.
+ *
+ * @param      da
+ * @param[in]  index   first item shifted
+ * @param[in]  offset  shift in no. of items
+ */
+void _array_shift(array_t *da, size_t index, size_t offset)
```

Comment:
I don't think this function needs to be considered "internal", and the underscore naming should be removed.  Also, it should reserve capacity if necessary instead of adding failure modes.